File: voc/store/weekly.py
"""The weekly customer review: one week of contacts answered as four questions.

Assembled from the same queries the tools and the dashboard run, so every number here is one a
reader can open. The four sections are deliberately in the order a reader needs them:

  1. what changed          volume and mix against the week before and the four before that
  2. what drove feeling    the specific moments, in the customers' own words
  3. what needs attention  ranked by growth and by reported impact, not by volume alone
  4. what the calls say    verbatim, with distinct call counts and what the evidence cannot bear

Section 4 states the limits out loud because a weekly review is read quickly and acted on, which is
exactly when an unstated caveat does damage.
"""
from __future__ import annotations

import sqlite3
from typing import Any

from voc.analytics.emerging import week_add
from voc.schemas.filters import Filters
from voc.store import queries as Q

RECENT_WEEKS = 4          # the comparison window behind the reviewed week
TOP_REASONS = 6
TOP_DRIVERS = 5
MAX_QUOTES = 6
NEW_THEME_LOOKBACK = 8    # a theme absent this many weeks before counts as new to the week


def _week_filters(base: Filters, week: str) -> Filters:
    """`base` narrowed to one ISO week."""
    monday = Q.week_monday(week)
    sunday = Q.week_sunday(week)
    return base.model_copy(update={"date_from": monday, "date_to": sunday})


def _range_filters(base: Filters, first_week: str, last_week: str) -> Filters:
    return base.model_copy(update={"date_from": Q.week_monday(first_week),
                                   "date_to": Q.week_sunday(last_week)})


def _delta(now: float | None, before: float | None) -> dict[str, Any]:
    """Change expressed both ways, with direction, or unknown when there is nothing to compare."""
    if now is None or before is None:
        return {"direction": "unknown", "points": None, "ratio": None}
    points = round(now - before, 1)
    ratio = round(now / before, 2) if before else None
    direction = "up" if points > 0.2 else "down" if points < -0.2 else "flat"
    return {"direction": direction, "points": points, "ratio": ratio}


def _reason_map(payload: dict[str, Any]) -> dict[str, dict[str, Any]]:
    return {r["reason"]: r for r in payload.get("rows", [])}


def what_changed(con: sqlite3.Connection, base: Filters, week: str) -> dict[str, Any]:
    """Section 1: volume, the reason mix, and what is new to this week."""
    this_week = _week_filters(base, week)
    prev_week = _week_filters(base, week_add(week, -1))
    prior = _range_filters(base, week_add(week, -RECENT_WEEKS), week_add(week, -1))

    now = Q.contact_reasons(con, this_week, compare_with_previous=False)
    before = Q.contact_reasons(con, prev_week, compare_with_previous=False)
    baseline = Q.contact_reasons(con, prior, compare_with_previous=False)

    n_now = now["scope"]["n_calls_in_scope"]
    n_before = before["scope"]["n_calls_in_scope"]
    n_baseline = baseline["scope"]["n_calls_in_scope"]
    weekly_average = round(n_baseline / RECENT_WEEKS, 1) if n_baseline else 0.0

    before_by = _reason_map(before)
    baseline_by = _reason_map(baseline)
    rows = []
    for r in now.get("rows", [])[:TOP_REASONS]:
        prev = before_by.get(r["reason"])
        base_row = baseline_by.get(r["reason"])
        base_share = round(100 * base_row["n_calls"] / n_baseline, 1) if base_row and n_baseline else None
        rows.append({
            "reason": r["reason"], "label": r["label"],
            "n_calls": r["n_calls"], "share_pct": r["share_pct"],
            "n_previous_week": prev["n_calls"] if prev else 0,
            "vs_previous_week": _delta(r["share_pct"], prev["share_pct"] if prev else None),
            "vs_four_week_average": _delta(r["share_pct"], base_share),
            "top_specific": [s["text"] for s in (r.get("top_specific_reasons") or [])[:2]],
            "call_ids": (r.get("sample_call_ids") or [])[:20],
        })

    # New to the week: a theme with calls now and none in the weeks just before it.
    seen_before = {t["theme_id"] for t in Q.list_themes(
        con, _range_filters(base, week_add(week, -NEW_THEME_LOOKBACK), week_add(week, -1)),
        sort_by="n_calls", limit=500).get("rows", [])}
    new_rows = [t for t in Q.list_themes(con, this_week, sort_by="n_calls", limit=500).get("rows", [])
                if t["theme_id"] not in seen_before]

    return {
        "week": week, "week_start": Q.week_monday(week), "week_end": Q.week_sunday(week),
        "n_calls": n_now, "n_calls_previous_week": n_before,
        "four_week_average": weekly_average,
        "vs_previous_week": _delta(float(n_now), float(n_before)),
        "vs_four_week_average": _delta(float(n_now), weekly_average),
        "reasons": rows,
        "new_this_week": [{"theme_id": t["theme_id"], "name": t["name"], "n_calls": t["n_calls"],
                           "driver_category": t.get("driver_category")} for t in new_rows[:5]],
        "n_new_this_week": len(new_rows),
    }


def what_drove_feeling(con: sqlite3.Connection, base: Filters, week: str) -> dict[str, Any]:
    """Section 2: the specific moments behind frustration and satisfaction, in the customers' words."""
    scope = _week_filters(base, week)
    out: dict[str, Any] = {"week": week}
    for polarity in ("negative", "positive"):
        payload = Q.sentiment_drivers(con, polarity, "theme", scope, limit=TOP_DRIVERS)
        rows = []
        for r in payload.get("rows", []):
            key = r.get("key")
            rows.append({
                "key": key, "name": r.get("name") or key,
                "theme_id": key if str(key).startswith("thm_") else None,
                "n_calls": r.get("n_calls"), "mean_sentiment": r.get("mean_sentiment"),
                "root_cause": r.get("root_cause"),
                "triggers": [t["text"] for t in (r.get("top_specific_drivers") or [])[:3]],
                "quotes": [{"quote": q["quote"], "call_id": q["call_id"],
                            "evidence_id": q.get("evidence_id")}
                           for q in (r.get("quotes") or [])[:2]],
            })
        out[polarity] = rows
        out[f"n_{polarity}_calls"] = sum(r["n_calls"] or 0 for r in rows)
    # On a complaint corpus positive topics rarely clear support; the moments carry it instead.
    moments = Q.sentiment_drivers(con, "positive", "driver_category", scope, limit=TOP_DRIVERS)
    out["positive_moments"] = (moments.get("data") or {}).get("positive_moments_by_category") or []
    return out


def needs_attention(con: sqlite3.Connection, base: Filters, week: str) -> dict[str, Any]:
    """Section 3: ranked by growth and by what customers say it cost them, not by volume alone."""
    emerging = Q.emerging_themes(con, week, base, min_recent=Q.MIN_SUPPORT, only_new=False, limit=8)
    scope = _week_filters(base, week)
    rows = []
    for r in emerging.get("rows", []):
        detail = Q.theme_detail(con, r["theme_id"], scope)
        data = detail.get("data") or {}
        rows.append({
            "theme_id": r["theme_id"], "name": r.get("name"), "status": r.get("status"),
            "n_recent": r.get("n_recent"), "expected_recent": r.get("expected_recent"),
            "z": r.get("z"), "robust_8w": r.get("robust_8w"),
            "first_seen_week": r.get("first_seen_week"),
            "n_calls_this_week": data.get("n_calls"),
            "unresolved_share": (data.get("resolution_split") or {}).get("unresolved_share"),
            "customer_asks": (data.get("customer_ask_split") or [])[:2],
            "root_cause": data.get("root_cause"),
            "call_ids": (detail.get("call_ids") or [])[:20],
        })
    data = emerging.get("data") or {}
    return {
        "week": week, "rows": rows,
        "n_tested": data.get("n_tested"),
        "expected_false_positives": data.get("expected_false_positives"),
        "status_counts": data.get("status_counts") or {},
        "fading": data.get("fading") or [],
    }


def supporting_calls(con: sqlite3.Connection, base: Filters, week: str,
                     sections: dict[str, Any]) -> dict[str, Any]:
    """Section 4: verbatim excerpts, distinct call counts, and what the evidence cannot bear."""
    scope = _week_filters(base, week)
    wanted: list[tuple[str, str]] = []
    for row in sections["needs_attention"]["rows"][:2]:
        wanted.append((row["theme_id"], row.get("name") or row["theme_id"]))
    for row in sections["drove_feeling"].get("negative", [])[:2]:
        if row.get("theme_id") and row["theme_id"] not in {w[0] for w in wanted}:
            wanted.append((row["theme_id"], row.get("name") or row["theme_id"]))

    recent = _range_filters(base, week_add(week, -(RECENT_WEEKS - 1)), week)
    groups = []
    for theme_id, name in wanted[:3]:
        window, used = week, scope
        quotes = Q.get_quotes(con, "theme", theme_id, None, used, n=MAX_QUOTES,
                              polarity="any", diverse=True)
        rows = quotes.get("rows", [])
        if len(rows) < 2:
            # One week is often too thin to read. Widen, and name the window rather than pretend.
            window, used = f"{week_add(week, -(RECENT_WEEKS - 1))}..{week}", recent
            quotes = Q.get_quotes(con, "theme", theme_id, None, used, n=MAX_QUOTES,
                                  polarity="any", diverse=True)
            rows = quotes.get("rows", [])
        groups.append({
            "theme_id": theme_id, "name": name, "window": window,
            "n_distinct_calls": len({r["call_id"] for r in rows}),
            "n_calls_in_window": quotes["scope"].get("n_calls_in_scope"),
            "quotes": [{"quote": r["quote"], "call_id": r["call_id"], "date": r.get("date"),
                        "product": r.get("product"), "evidence_id": r.get("evidence_id")}
                       for r in rows],
        })

    n_week = sections["changed"]["n_calls"]
    limits = [
        f"Every count here is contacts in {week} only, and one contact counts once per theme.",
        f"Nothing is ranked below {Q.MIN_SUPPORT} contacts, so a real problem seen {Q.MIN_SUPPORT - 1} "
        f"times this week is deliberately absent from these lists.",
        "Quotes are exact substrings of what the customer wrote or said; anything that did not verify "
        "was dropped rather than paraphrased.",
    ]
    widened = [g["name"] for g in groups if g["window"] != week]
    if widened:
        limits.append(f"{len(widened)} of these groups had fewer than two verified quotes inside the "
                      f"week, so their excerpts are drawn from the last {RECENT_WEEKS} weeks; each "
                      f"group states the window it used.")
    if n_week and n_week < 40:
        limits.append(f"Only {n_week} contacts landed in this week, so week-on-week movement here is "
                      f"noisy and a single day's intake can swing it.")
    return {"week": week, "groups": groups, "limits": limits}


def weekly_review(con: sqlite3.Connection, week: str | None, filters: Filters | None = None) -> dict[str, Any]:
    """All four sections for one week. `week` defaults to the index's as-of week."""
    base = filters or Filters()
    week = week or Q.as_of_or_meta(con, None)
    if not week:
        raise Q.QueryError("no week available: the index has no as_of_week")
    if week not in set(Q.available_weeks(con)):
        raise Q.NotFound(f"no contacts in {week}")
    changed = what_changed(con, base, week)
    drove = what_drove_feeling(con, base, week)
    attention = needs_attention(con, base, week)
    sections = {"changed": changed, "drove_feeling": drove, "needs_attention": attention}
    sections["supporting"] = supporting_calls(con, base, week, sections)
    return {
        "week": week,
        "week_start": changed["week_start"], "week_end": changed["week_end"],
        "filters": base.canonical(),
        "sections": sections,
    }

File: voc/web/js/weekly.js
// The weekly customer review: four questions about one week, each answered from the same
// queries the dashboard runs. Every count is a link to the calls behind it.
import { api } from "./api.js";
import { openCall, openCallList } from "./calldrawer.js";
import { clear, el, label, num, pctOf, plural } from "./format.js";
import { openTheme } from "./themecard.js";

const QUESTIONS = [
  ["changed", "What changed in customer contacts this week?",
   "Top reasons, new issues and changes in share of calls."],
  ["drove_feeling", "Which experiences drove frustration or satisfaction?",
   "Specific moments and recurring causes in customers' own words."],
  ["needs_attention", "Which issues need attention first?",
   "Growing pain points, call volumes and reported customer impact."],
  ["supporting", "What do the supporting calls actually say?",
   "Exact excerpts, distinct call counts and limits of the evidence."],
];

let weeks = [];
let current = null;

const arrow = (d) => (d === "up" ? "▲" : d === "down" ? "▼" : d === "flat" ? "→" : "");
const moveClass = (d) => (d === "up" ? "up" : d === "down" ? "down" : "flat");

function movement(delta, suffix = " pts") {
  if (!delta || delta.direction === "unknown" || delta.points === null) {
    return el("span", { class: "move flat", text: "no prior week" });
  }
  const sign = delta.points > 0 ? "+" : "";
  return el("span", { class: `move ${moveClass(delta.direction)}`,
                      text: `${arrow(delta.direction)} ${sign}${delta.points}${suffix}` });
}

function quoteLine(q) {
  return el("p", { class: "quote" }, [
    el("button", { class: "linkish", text: `"${q.quote}"`,
                   onclick: () => openCall(q.call_id, q.quote) }),
  ]);
}

/** Section 1: volume and mix. */
function sectionChanged(d) {
  const out = el("div", {});
  out.appendChild(el("div", { class: "weekly-stats" }, [
    stat(num(d.n_calls), "contacts this week"),
    stat(num(d.n_calls_previous_week), "the week before"),
    stat(String(d.four_week_average), "four-week average"),
    stat(String(d.n_new_this_week), "themes new this week"),
  ]));
  const versus = el("p", { class: "footnote" }, [
    document.createTextNode("Against the four-week average: "), movement(d.vs_four_week_average, ""),
  ]);
  out.appendChild(versus);

  for (const r of d.reasons || []) {
    out.appendChild(el("div", { class: "row" }, [
      el("div", { class: "row-head" }, [
        el("span", { class: "row-name", text: r.label }),
        el("span", { class: "row-meta" }, [
          document.createTextNode(`${num(r.n_calls)} · ${pctOf(r.share_pct)}  `),
          movement(r.vs_previous_week),
        ]),
      ]),
      r.top_specific?.length ? el("p", { class: "quote", text: r.top_specific[0] }) : null,
    ]));
  }
  if (d.new_this_week?.length) {
    out.appendChild(el("h3", { class: "weekly-sub", text: "New to this week" }));
    for (const t of d.new_this_week) {
      out.appendChild(el("div", { class: "row" }, [
        el("div", { class: "row-head" }, [
          el("button", { class: "linkish row-name", text: t.name, onclick: () => openTheme(t.theme_id) }),
          el("span", { class: "row-meta", text: plural(t.n_calls, "call") }),
        ]),
      ]));
    }
  }
  return out;
}

/** Section 2: what drove frustration and satisfaction. */
function sectionFeeling(d) {
  const out = el("div", {});
  for (const [key, title] of [["negative", "Frustration"], ["positive", "Satisfaction"]]) {
    const rows = d[key] || [];
    out.appendChild(el("h3", { class: "weekly-sub", text: title }));
    if (!rows.length && key === "positive" && (d.positive_moments || []).length) {
      for (const m of d.positive_moments) {
        out.appendChild(el("div", { class: "row" }, [
          el("div", { class: "row-head" }, [
            el("span", { class: "row-name", text: label(m.category) }),
            el("span", { class: "row-meta", text: plural(m.n_calls, "call") }),
          ]),
          (m.quotes || [])[0] ? quoteLine(m.quotes[0]) : null,
        ]));
      }
      continue;
    }
    if (!rows.length) {
      out.appendChild(el("p", { class: "muted", text: "nothing above minimum support this week" }));
      continue;
    }
    for (const r of rows) {
      out.appendChild(el("div", { class: "row" }, [
        el("div", { class: "row-head" }, [
          r.theme_id
            ? el("button", { class: "linkish row-name", text: r.name, onclick: () => openTheme(r.theme_id) })
            : el("span", { class: "row-name", text: r.name }),
          el("span", { class: "row-meta", text: `${plural(r.n_calls, "call")} · mean ${r.mean_sentiment}` }),
        ]),
        r.triggers?.length ? el("p", { class: "trigger", text: r.triggers[0] }) : null,
        (r.quotes || [])[0] ? quoteLine(r.quotes[0]) : null,
      ]));
    }
  }
  return out;
}

/** Section 3: what to act on first. */
function sectionAttention(d) {
  const out = el("div", {});
  if (!d.rows?.length) {
    out.appendChild(el("p", { class: "muted",
      text: `Nothing cleared the threshold this week. ${num(d.n_tested)} themes were testable.` }));
  }
  for (const r of d.rows || []) {
    const asks = (r.customer_asks || []).map((a) => `${label(a.value)} ${pctOf(a.share_pct)}`).join(" · ");
    out.appendChild(el("div", { class: "row" }, [
      el("div", { class: "row-head" }, [
        el("button", { class: "linkish row-name", text: r.name, onclick: () => openTheme(r.theme_id) }),
        el("span", { class: "pill " + (r.status || ""), text: (r.status || "").toUpperCase() }),
      ]),
      el("p", { class: "footnote",
        text: `${num(r.n_recent)} in the last four weeks against ${Number(r.expected_recent).toFixed(1)} expected`
              + ` · z ${Number(r.z).toFixed(1)}`
              + (r.robust_8w ? " · robust at 8 weeks" : "")
              + (r.first_seen_week ? ` · first seen ${r.first_seen_week}` : "") }),
      r.root_cause ? el("p", { class: "trigger", text: r.root_cause }) : null,
      asks ? el("p", { class: "footnote", text: `What they asked for: ${asks}` }) : null,
    ]));
  }
  if (d.expected_false_positives !== undefined && d.expected_false_positives !== null) {
    out.appendChild(el("p", { class: "footnote",
      text: `${Number(d.expected_false_positives).toFixed(2)} of ${num(d.n_tested)} themes tested would`
            + ` pass this threshold by chance.` }));
  }
  return out;
}

/** Section 4: the words themselves, and what they cannot bear. */
function sectionSupporting(d) {
  const out = el("div", {});
  for (const g of d.groups || []) {
    out.appendChild(el("div", { class: "row" }, [
      el("div", { class: "row-head" }, [
        el("button", { class: "linkish row-name", text: g.name, onclick: () => openTheme(g.theme_id) }),
        el("span", { class: "row-meta",
          text: `${plural(g.n_distinct_calls, "distinct call")}${g.window && g.window.includes("..") ? ` · ${g.window}` : ""}` }),
      ]),
      ...(g.quotes || []).map(quoteLine),
    ]));
  }
  if (d.limits?.length) {
    out.appendChild(el("h3", { class: "weekly-sub", text: "What this evidence cannot bear" }));
    const list = el("ul", { class: "limits" });
    for (const l of d.limits) list.appendChild(el("li", { text: l }));
    out.appendChild(list);
  }
  return out;
}

const RENDER = { changed: sectionChanged, drove_feeling: sectionFeeling,
                 needs_attention: sectionAttention, supporting: sectionSupporting };

function stat(value, caption) {
  return el("div", { class: "weekly-stat" }, [
    el("strong", { text: value }), el("span", { text: caption }),
  ]);
}

async function render(week) {
  const body = document.getElementById("weeklyBody");
  clear(body).appendChild(el("p", { class: "muted", text: "reading the week…" }));
  let payload;
  try {
    payload = await api.weekly(week);
  } catch (err) {
    clear(body).appendChild(el("p", { class: "muted", text: `could not load: ${err.message}` }));
    return;
  }
  current = payload.week;
  document.getElementById("weeklyRange").textContent =
    `${payload.week} · ${payload.week_start} to ${payload.week_end}`;
  const pick = document.getElementById("weekPick");
  if (pick.value !== current) pick.value = current;

  const out = clear(body);
  for (const [key, question, blurb] of QUESTIONS) {
    const section = payload.sections[key];
    out.appendChild(el("article", { class: "weekly-section" }, [
      el("h2", { text: question }),
      el("p", { class: "muted", text: blurb }),
      section ? RENDER[key](section) : el("p", { class: "muted", text: "unavailable" }),
    ]));
  }
}

function step(delta) {
  const i = weeks.indexOf(current);
  if (i < 0) return;
  const next = weeks[i + delta];
  if (next) render(next);
}

export async function initWeekly(defaultWeek) {
  const panel = document.getElementById("weekly");
  const pick = document.getElementById("weekPick");
  document.getElementById("weeklyBtn").addEventListener("click", async () => {
    panel.hidden = false;
    document.querySelector("main").hidden = true;
    if (!weeks.length) {
      weeks = (await api.weeks()).weeks || [];
      clear(pick);
      for (const w of weeks) pick.appendChild(el("option", { value: w, text: w }));
    }
    render(current || defaultWeek || weeks[weeks.length - 1]);
  });
  document.getElementById("weeklyClose").addEventListener("click", () => {
    panel.hidden = true;
    document.querySelector("main").hidden = false;
  });
  pick.addEventListener("change", () => render(pick.value));
  document.getElementById("weekPrev").addEventListener("click", () => step(-1));
  document.getElementById("weekNext").addEventListener("click", () => step(1));
}
